PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+                break
+            elif lineage_clustering[isolate2] == null_cluster_value:
+                seed_isolate = isolate2
+                break
+    # return identified seed isolate
+    return seed_isolate
+
+def get_lineage_clustering_information(seed_isolate, row_labels, distances):
+    """ Generates the ranked distances needed for cluster
+    definition.
+    
+    Args:
+        seed_isolate (str)
+            Isolate used to initiate lineage.
+        row_labels (list of tuples)
+            Pairs of isolates labelling each distance.
+        distances (numpy array)
+            Pairwise distances used for defining relationships.
+    
+    Returns:
+        lineage_info (dict)
+            Dict listing isolates by ranked distance from seed.
+    
+    """
+    # data structure
+    lineage_info = defaultdict(list)
+    rank = 0
+    last_dist = -1
+    # iterate through distances recording rank
+    for index,distance in enumerate(distances):
+        if seed_isolate in row_labels[index]:
+            if distance > last_dist:
+                rank = rank + 1
+                last_dist = distance
+            pair = row_labels[index][0] if row_labels[index][1] == seed_isolate else row_labels[index][1]
+            lineage_info[rank].append(pair)
+    # return information
+    return lineage_info
+
+def generate_nearest_neighbours(distances, row_labels, isolate_list, R):
+    # data structures
+    nn = {}
+    last_dist = {}
+    num_ranks = {}
+    for i in isolate_list:
+        nn[i] = {}
+        num_ranks[i] = 0
+    total_isolates = len(isolate_list)
+    num_distances = len(distances)
+    completed_isolates = 0
+    index = 0
+    # iterate through distances until all nearest neighbours identified
+    while completed_isolates < total_isolates and index < num_distances:
+        distance = distances[index]
+        # iterate through all listed isolates
+        for isolate in row_labels[index]:
+            if isolate in num_ranks.keys() and num_ranks[isolate] < R:
+                # R is the number of unique distances so only increment if
+                # different from the last distance
+                if isolate in last_dist.keys() and last_dist[isolate] != distance:
+                    num_ranks[isolate] = num_ranks[isolate] + 1
+                # if maximum number of ranks reached, mark as complete
+                if num_ranks[isolate] == R: # stop at R as counting from 0
+                    completed_isolates = completed_isolates + 1
+                # if not, add as a nearest neighbour
+                else:
+                    pair = row_labels[index][0] if row_labels[index][1] == isolate else row_labels[index][1]
+                    nn[isolate][pair] = distance
+                last_dist[isolate] = distance
+        index = index + 1
+    # return completed dict
+    return nn
+
+def update_nearest_neighbours(distances, row_labels, R, qlist, nn, lineage_clustering, null_cluster_value):
```

Comment:
This could be replaced with a call to generate nearest neighbours, by first constructing the full matrix